docs/protobuf-libraries.md

```diff
@@ -0,0 +1,115 @@
+Loren & Roey surveyed the available protobuf libraries in Dec '21 for use with our `ProtobufBinaryDataConverter` and `ProtobufJsonDataConverter`. The main criteria was:
+
+- A. TypeScript types for messages
+- B. Being able to check at runtime whether an object passed to the SDK as input or returned to the SDK from a workflow/query/activity is meant to be protobuf-serialized, without adding annotations to the functions.
+- C. Spec-compliant [proto3 JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json) so that the TS SDK is interoperable with the other SDKs
+
+## Options
+
+### protobufjs
+
+A and B, but not C.
+
+- Most popular lib (5M downloads/wk)
+- Fairly inactive maintainers (infrequent updates, many open PRs & issues)
+- Non-standard JSON serialization
```

Comment:
What makes it non-standard? I expect every message class to have a [toJSON](https://protobufjs.github.io/protobuf.js/Message.html#toJSON) method and a `fromJSON` static method.

Reply:
Every class has a `toJSON` method, but `JSON.stringify()` doesn't match the spec:

https://developers.google.com/protocol-buffers/docs/proto3#json

Eg `stringify` throws on BigInts and doesn't base64-encode binary. protobufjs hasn't implemented a number of the types that need custom encoding: https://github.com/protobufjs/protobuf.js/issues/1304

Reply:
> Eg stringify throws on BigInts and doesn't base64-encode binary.

Why use stringify? Doesn't the toJSON directly work with these? Is there an issue to follow?

> protobufjs hasn't implemented a number of the types that need custom encoding: https://github.com/protobufjs/protobuf.js/issues/1304

Makes sense. They just need better support for well known types. I wonder if we can contribute? Otherwise, I wonder if we can walk the object for WKT's and convert them ourselves.

Reply:
The last commit in their repo was in last May, I fear contribution is not a viable option.

Reply:
> Why use stringify? Doesn't the toJSON directly work with these? Is there an issue to follow?

I just meant that if `toJSON` were implemented with `JSON.stringify`, it wouldn't work. I made this issue: https://github.com/temporalio/sdk-typescript/issues/481

> I wonder if we can contribute? Otherwise, I wonder if we can walk the object for WKT's and convert them ourselves.

I'd like to try contributing. If converting ourselves, unless we fork protobufjs, we'd be doing what `proto3-json-serializer` does, and would have the same requirement on how the user generates & provides their messages.

Reply:
:+1: I am not sure it's reasonable to ask people to use our custom protoc. Ideally we can t do this from the outside on protobufjs-generated classes/descriptors.

Having said that, this being the only one of our languages without a canonical, supported proto-JSON implementation, I'd support just erroring in data converters if protos are provided in either direction, but allow them to use a custom converter to pick their preferred conversion (or binary only, but we should still error on from-payload when we receive `json/protobuf`).

Reply:
> to ask people to use our custom protoc

We're still having them use the protobufjs generator (`pbjs`). It may not be the way they're currently running it. If not, I think it's straightforward to switch or to do both—our way for temporal code and their way for the rest of their code. The main difference in use is doing `Message.create()` instead of `new Message()`.

If we think running the generator our way is too big an ask for some users, we could add support for just `json/binary` for them, without removing current functionality.

> the only one of our languages without a canonical, supported proto-JSON implementation

I think the library we're using is well-maintained and from Google/canonical: https://github.com/googleapis/proto3-json-serializer-nodejs

Reply:
> Can you document why in one of your docs? Is this a limitation of proto3-json-serializer?

Yes. Limitation of this way of running `pbjs`.